File: drivers/python-asyncio/src/reddb_asyncio/http.py
"""HTTP/HTTPS transport for the asyncio driver.

The RedDB HTTP server exposes a REST surface that mirrors the
RedWire data-plane (see ``src/server/routing.rs``). This module
keeps the same public method names as :class:`RedwireClient` so the
top-level ``Reddb`` facade can route either way without callers
caring.

The primary endpoint mapping matches what the JS driver speaks
(``drivers/js/src/http.js``):

==========================  =================================================
Method                      Endpoint
==========================  =================================================
``query``                   ``POST /query`` (body ``{"query": sql}``)
``insert``                  ``POST /collections/{name}/rows``
``bulk_insert``             ``POST /collections/{name}/bulk/rows``
``scan``                    ``GET  /collections/{name}/scan?limit=...``
``get``                     ``GET  /collections/{name}/{id}``
``delete``                  ``DELETE /collections/{name}/{id}``
``kv_put``                  ``PUT  /collections/{name}/kv/{key}`` (fallback ``/kvs/{key}``)
``kv_get``                  ``GET  /collections/{name}/kv/{key}`` (fallback ``/kvs/{key}``)
``kv_delete``               ``DELETE /collections/{name}/kv/{key}`` (fallback ``/kvs/{key}``)
``kv_watch``                ``GET  /collections/{name}/kv/{key}/watch`` (SSE)
``ping`` / health           ``GET  /health``
``login``                   ``POST /auth/login``
==========================  =================================================
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Any
from urllib.parse import quote

import httpx

from .errors import EngineError, HttpError

DEFAULT_TIMEOUT_SECS: float = 30.0


class HttpClient:
    """Thin async HTTP client matching the RedWire surface.

    ``base_url`` should NOT include a trailing slash; we strip it just
    in case. Pass ``token`` to attach ``Authorization: Bearer ...`` to
    every request.
    """

    def __init__(
        self,
        *,
        base_url: str,
        token: str | None = None,
        timeout: float = DEFAULT_TIMEOUT_SECS,
        verify: bool | str = True,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self._token = token
        self._owns_client = client is None
        if client is None:
            client = httpx.AsyncClient(
                base_url=self.base_url,
                timeout=timeout,
                verify=verify,
                http2=False,
                headers={"accept": "application/json"},
            )
        self._client = client

    @property
    def token(self) -> str | None:
        return self._token

    def set_token(self, token: str | None) -> None:
        self._token = token

    async def close(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    # ------------------------------------------------------------------ auth

    async def login(self, username: str, password: str) -> dict[str, Any]:
        """POST /auth/login → returns the parsed response. Stores the token
        on the client so subsequent calls carry it automatically.
        """
        body = await self._post_json("/auth/login", {"username": username, "password": password})
        token = body.get("token") if isinstance(body, dict) else None
        if isinstance(token, str):
            self.set_token(token)
        return body if isinstance(body, dict) else {"raw": body}

    # ------------------------------------------------------------------ ops

    async def query(self, sql: str) -> dict[str, Any]:
        return await self._post_json("/query", {"query": sql})

    async def insert(self, collection: str, payload: dict[str, Any]) -> dict[str, Any]:
        return await self._post_json(f"/collections/{quote(collection)}/rows", payload)

    async def bulk_insert(self, collection: str, payloads: list[dict[str, Any]]) -> dict[str, Any]:
        return await self._post_json(
            f"/collections/{quote(collection)}/bulk/rows", {"rows": payloads}
        )

    async def get(self, collection: str, doc_id: str) -> dict[str, Any]:
        url = f"/collections/{quote(collection)}/{quote(str(doc_id))}"
        return await self._request("GET", url)

    async def delete(self, collection: str, doc_id: str) -> dict[str, Any]:
        url = f"/collections/{quote(collection)}/{quote(str(doc_id))}"
        return await self._request("DELETE", url)

    async def kv_put(self, collection: str, key: str, value: Any) -> dict[str, Any]:
        return await self._request_with_legacy_kv(
            "PUT",
            collection,
            key,
            json_body={"value": value},
        )

    async def kv_get(self, collection: str, key: str) -> dict[str, Any]:
        return await self._request_with_legacy_kv("GET", collection, key)

    async def kv_delete(self, collection: str, key: str) -> dict[str, Any]:
        return await self._request_with_legacy_kv("DELETE", collection, key)

    async def kv_incr(
        self, collection: str, key: str, by: int = 1, ttl_ms: int | None = None
    ) -> dict[str, Any]:
        return await self._request_kv_counter("incr", collection, key, by, ttl_ms)

    async def kv_decr(
        self, collection: str, key: str, by: int = 1, ttl_ms: int | None = None
    ) -> dict[str, Any]:
        return await self._request_kv_counter("decr", collection, key, by, ttl_ms)

    async def kv_watch(self, collection: str, key: str) -> AsyncIterator[Any]:
        """Watch a single KV key via HTTP SSE.

        Yields parsed JSON ``data:`` payloads when possible, otherwise
        yields raw data strings.
        """

        collection_q = quote(collection)
        key_q = quote(str(key))
        headers = {**self._headers(), "accept": "text/event-stream"}
        try:
            async with self._client.stream(
                "GET",
                f"/collections/{collection_q}/kv/{key_q}/watch",
                headers=headers,
            ) as response:
                if response.is_error:
                    await response.aread()
                    _parse_response(response)
                data: list[str] = []
                async for line in response.aiter_lines():
                    event = _consume_sse_line(line, data)
                    data = event["data"]
                    if event["emit"]:
                        yield event["value"]
                if data:
                    yield _parse_sse_data(data)
        except httpx.HTTPError as exc:  # pragma: no cover - network
            raise HttpError(str(exc), status=0, body="") from exc

    async def scan(self, collection: str, **params: Any) -> dict[str, Any]:
        url = f"/collections/{quote(collection)}/scan"
        return await self._request("GET", url, params=params or None)

    async def ping(self) -> dict[str, Any]:
        return await self._request("GET", "/health")

    async def health(self) -> dict[str, Any]:  # alias
        return await self.ping()

    # ----------------------------------------------------------------- low-

    def _headers(self) -> dict[str, str]:
        h: dict[str, str] = {}
        if self._token:
            h["authorization"] = f"Bearer {self._token}"
        return h

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json_body: Any | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        try:
            response = await self._client.request(
                method,
                path,
                json=json_body,
                params=params,
                headers=self._headers(),
            )
        except httpx.HTTPError as exc:  # pragma: no cover - network
            raise HttpError(str(exc), status=0, body="") from exc
        return _parse_response(response)

    async def _request_with_legacy_kv(
        self,
        method: str,
        collection: str,
        key: str,
        *,
        json_body: Any | None = None,
    ) -> Any:
        collection_q = quote(collection)
        key_q = quote(str(key))
        canonical = f"/collections/{collection_q}/kv/{key_q}"
        legacy = f"/collections/{collection_q}/kvs/{key_q}"
        try:
            response = await self._client.request(
                method,
                canonical,
                json=json_body,
                headers=self._headers(),
            )
            if response.status_code == 404:
                response = await self._client.request(
                    method,
                    legacy,
                    json=json_body,
                    headers=self._headers(),
                )
        except httpx.HTTPError as exc:  # pragma: no cover - network
            raise HttpError(str(exc), status=0, body="") from exc
        return _parse_response(response)

    async def _request_kv_counter(
        self,
        op: str,
        collection: str,
        key: str,
        by: int,
        ttl_ms: int | None,
    ) -> Any:
        collection_q = quote(collection)
        key_q = quote(str(key))
        params: dict[str, Any] = {"by": by}
        if ttl_ms is not None:
            params["ttl_ms"] = ttl_ms
        return await self._request(
            "POST",
            f"/collections/{collection_q}/kv/{key_q}/{op}",
            params=params,
        )

    async def _post_json(self, path: str, body: Any) -> Any:
        headers = {**self._headers(), "content-type": "application/json"}
        try:
            response = await self._client.post(path, json=body, headers=headers)
        except httpx.HTTPError as exc:  # pragma: no cover - network
            raise HttpError(str(exc), status=0, body="") from exc
        return _parse_response(response)


def _parse_response(response: httpx.Response) -> Any:
    text = response.text
    parsed: Any = None
    if text:
        try:
            parsed = response.json()
        except Exception:
            parsed = {"raw": text}
    if response.is_error:
        code = (
            parsed.get("error_code") if isinstance(parsed, dict) else None
        ) or f"HTTP_{response.status_code}"
        message = (
            (parsed.get("error") if isinstance(parsed, dict) else None)
            or (parsed.get("message") if isinstance(parsed, dict) else None)
            or f"request failed with status {response.status_code}"
        )
        raise HttpError(message, status=response.status_code, body=text)
    if isinstance(parsed, dict) and "ok" in parsed:
        if parsed["ok"] is False:
            raise EngineError(
                parsed.get("error") or "RPC error",
                code=parsed.get("error_code", "RPC_ERROR"),
                payload=parsed,
            )
        return parsed.get("result", parsed)
    return parsed


def _consume_sse_line(line: str, data: list[str]) -> dict[str, Any]:
    if line == "":
        if not data:
            return {"emit": False, "data": data}
        return {"emit": True, "value": _parse_sse_data(data), "data": []}
    if line.startswith(":"):
        return {"emit": False, "data": data}
    field, sep, value = line.partition(":")
    if sep and value.startswith(" "):
        value = value[1:]
    if field == "data":
        data.append(value)
    return {"emit": False, "data": data}


def _parse_sse_data(lines: list[str]) -> Any:
    text = "\n".join(lines)
    if text == "":
        return None
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return text


__all__ = ["HttpClient", "DEFAULT_TIMEOUT_SECS"]

File: drivers/js-client/src/http.js
/**
 * HTTP transport for the JS driver.
 *
 * Mirrors the public surface of `RpcClient` (call, close) but talks
 * straight to the RedDB HTTP server via fetch() — no binary spawn.
 * Each `RedDB` method is mapped to a REST endpoint; method names
 * stay identical to the JSON-RPC ones so `RedDB` doesn't need to
 * know which transport it's using.
 *
 * Endpoint mapping (server-side defined in src/server/routing.rs):
 *
 *   query / explain     → POST /query, POST /query/explain
 *   insert              → POST /collections/:name/rows
 *   bulk_insert         → POST /collections/:name/bulk/rows
 *   get                 → GET  /collections/:name/{id}      (entity scan + filter)
 *   delete              → DELETE /collections/:name/{id}
 *   health              → GET  /health
 *   version             → GET  /admin/version
 *   auth.login          → POST /auth/login
 *   auth.whoami         → GET  /auth/whoami
 *   auth.create_api_key → POST /auth/api-keys
 *   auth.revoke_api_key → DELETE /auth/api-keys/:key
 *   auth.change_password→ POST /auth/change-password
 *
 *   cache.get               → GET    /cache/ns/:ns/:key
 *   cache.put               → PUT    /cache/ns/:ns/:key
 *   cache.exists            → GET    /cache/ns/:ns/:key/exists
 *   cache.invalidate        → DELETE /cache/ns/:ns/:key
 *   cache.invalidate_prefix → DELETE /cache/ns/:ns?prefix=...
 *   cache.invalidate_tags   → DELETE /cache/ns/:ns/tags  (body: {tags})
 *   cache.flush_namespace   → POST   /admin/blob_cache/flush_namespace
 *
 * Auth: every request after construction carries `Authorization:
 * Bearer <token>` (when set). `setToken(token)` updates it in
 * place — used by the login flow that exchanges credentials for
 * a fresh bearer.
 */

import { RedDBError } from './protocol.js'
import { classifyNdjsonFrame, splitLines } from './core/ndjson.js'

export class HttpRpcClient {
  /**
   * @param {object} opts
   * @param {string} opts.baseUrl   Server origin, e.g. 'https://reddb.example.com:55555'
   * @param {string} [opts.token]   Bearer token / API key
   */
  constructor({ baseUrl, token }) {
    if (typeof baseUrl !== 'string' || baseUrl.length === 0) {
      throw new TypeError('HttpRpcClient: baseUrl required')
    }
    this.baseUrl = baseUrl.replace(/\/$/, '')
    this.token = token ?? null
  }

  setToken(token) {
    this.token = token
  }

  async close() {
    // HTTP is stateless — nothing to close.
  }

  /**
   * Generic RPC entry point. Routes the named method to the
   * corresponding HTTP endpoint and returns the parsed JSON body.
   */
  async call(method, params = {}) {
    const route = ROUTES[method]
    if (!route) {
      throw new RedDBError(
        'UNKNOWN_METHOD',
        `HTTP transport has no route for method '${method}'`,
      )
    }
    const { url, init } = route(this.baseUrl, params)
    const response = await fetch(url, this.attachAuth(init))
    return parseResponse(response)
  }

  attachAuth(init) {
    const headers = new Headers(init.headers || {})
    if (this.token && !headers.has('authorization')) {
      headers.set('authorization', `Bearer ${this.token}`)
    }
    if (init.body && !headers.has('content-type')) {
      headers.set('content-type', 'application/json')
    }
    return { ...init, headers }
  }

  authHeaders(extra = {}) {
    const headers = new Headers(extra)
    if (this.token && !headers.has('authorization')) {
      headers.set('authorization', `Bearer ${this.token}`)
    }
    return headers
  }

  /**
   * Open a streaming read against `POST /query/stream`. Returns an async
   * iterable of typed frames (see streaming.js) plus a `cancel(reason)`
   * that aborts the underlying fetch. A non-streaming refusal (e.g. a
   * non-read-only statement) is surfaced as a rejected `RedDBError`
   * before any frame is yielded, so callers can tell "never accepted"
   * from a mid-stream failure.
   *
   * @param {{ sql?: string, cursor?: string, signal?: AbortSignal }} opts
   */
  async streamSelect({ sql, cursor, signal } = {}) {
    const controller = new AbortController()
    linkSignal(signal, controller)
    const body = cursor != null ? { cursor } : { query: sql }
    const response = await fetch(`${this.baseUrl}/query/stream`, {
      method: 'POST',
      headers: this.authHeaders({ 'content-type': 'application/json' }),
      body: JSON.stringify(body),
      signal: controller.signal,
    })
    if (!response.ok) {
      await throwHttpStreamRefusal(response)
    }
    const reader = response.body?.getReader()
    if (!reader) {
      throw new RedDBError('STREAM_PROTOCOL', 'streaming response had no body')
    }
    return {
      [Symbol.asyncIterator]() {
        return ndjsonFrameIterator(reader, controller)
      },
      async cancel() {
        controller.abort()
        try {
          await reader.cancel()
        } catch {
          // best-effort
        }
      },
    }
  }

  /**
   * Open a streaming write against `POST /streams/input`. The request
   * body is an NDJSON stream: an `open` frame (target + columns), then
   * one `row` frame per record. Backpressure flows through the request
   * body's writer. The terminal envelope is returned by `close()`.
   *
   * @param {{ target: string, columns?: string[], signal?: AbortSignal }} opts
   */
  async streamInput({ target, columns, signal } = {}) {
    const controller = new AbortController()
    linkSignal(signal, controller)
    const transform = new TransformStream()
    const writer = transform.writable.getWriter()
    const fetchPromise = fetch(`${this.baseUrl}/streams/input`, {
      method: 'POST',
      headers: this.authHeaders({ 'content-type': 'application/x-ndjson' }),
      body: transform.readable,
      duplex: 'half',
      signal: controller.signal,
    })
    // Surface a connection/refusal failure on the write path too, rather
    // than leaving the promise unhandled if the caller never calls close().
    fetchPromise.catch(() => {})

    let opened = false
    let cols = Array.isArray(columns) && columns.length > 0 ? columns.slice() : null
    const encodeLine = (obj) => writer.write(`${JSON.stringify(obj)}\n`)
    const ensureOpen = async (row) => {
      if (opened) return
      if (!cols) {
        cols = row && typeof row === 'object' ? Object.keys(row) : null
      }
      if (!cols || cols.length === 0) {
        throw new RedDBError(
          'INVALID_STREAM_COLUMNS',
          'inputStream() needs a non-empty column set — pass { columns } or write at least one object row',
        )
      }
      await writer.write(`${JSON.stringify({ open: { target, columns: cols } })}\n`)
      opened = true
    }

    return {
      async write(row) {
        await ensureOpen(row)
        await writer.ready
        await encodeLine({ row })
      },
      async close() {
        await ensureOpen(null)
        await writer.close()
        const response = await fetchPromise
        return await readInputTerminal(response)
      },
      async cancel(reason) {
        controller.abort()
        try {
          await writer.abort(reason)
        } catch {
          // best-effort — the abort above already tore the request down.
        }
      },
    }
  }
}

function linkSignal(signal, controller) {
  if (!signal) return
  if (signal.aborted) {
    controller.abort(signal.reason)
    return
  }
  signal.addEventListener('abort', () => controller.abort(signal.reason), { once: true })
}

async function throwHttpStreamRefusal(response) {
  const text = await response.text().catch(() => '')
  let body = null
  if (text) {
    try {
      body = JSON.parse(text)
    } catch {
      body = { raw: text }
    }
  }
  const code = body?.code || body?.error_code || `HTTP_${response.status}`
  const message =
    body?.error || body?.message || `stream refused with status ${response.status}`
  throw new RedDBError(code, message, body)
}

/** Async iterator over NDJSON frames from a web ReadableStream reader. */
async function* ndjsonFrameIterator(reader, controller) {
  const decoder = new TextDecoder()
  let buffer = ''
  try {
    for (;;) {
      const { value, done } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      const { lines, rest } = splitLines(buffer)
      buffer = rest
      for (const line of lines) {
        const frame = classifyNdjsonFrame(line)
        if (frame) yield frame
      }
    }
    const tail = buffer + decoder.decode()
    const frame = classifyNdjsonFrame(tail)
    if (frame) yield frame
  } finally {
    controller.abort()
  }
}

/** Read the input-stream response body and return its terminal envelope. */
async function readInputTerminal(response) {
  const text = await response.text()
  if (!response.ok) {
    let body = null
    try {
      body = text ? JSON.parse(text) : null
    } catch {
      body = { raw: text }
    }
    const code = body?.code || body?.error_code || `HTTP_${response.status}`
    const message = body?.error || body?.message || `input stream failed (${response.status})`
    throw new RedDBError(code, message, body)
  }
  const lines = text.split('\n').map((l) => l.trim()).filter((l) => l.length > 0)
  let end = null
  for (const line of lines) {
    const frame = classifyNdjsonFrame(line) // throws RedDBError on an {error} frame
    if (frame && frame.type === 'end') end = frame.value
  }
  if (!end) {
    throw new RedDBError('STREAM_PROTOCOL', 'input stream closed without a terminal end envelope')
  }
  return end
}

async function parseResponse(response) {
  const text = await response.text()
  let body = null
  if (text) {
    try {
      body = JSON.parse(text)
    } catch {
      body = { raw: text }
    }
  }
  if (!response.ok) {
    const code = body?.error_code || `HTTP_${response.status}`
    const message = body?.error || body?.message || `request failed with status ${response.status}`
    throw new RedDBError(code, message, body)
  }
  // RedDB envelope is `{ ok, result, error? }` for some endpoints
  // and bare JSON for others. Unwrap the envelope when present.
  if (body && typeof body === 'object' && 'ok' in body) {
    if (body.ok === false) {
      const code = body.error_code || 'RPC_ERROR'
      throw new RedDBError(code, body.error || 'unknown error', body)
    }
    return body.result ?? body
  }
  return body
}

// ---------------------------------------------------------------------------
// Method → HTTP route mapping
// ---------------------------------------------------------------------------

const ROUTES = {
  health: (base) => ({ url: `${base}/health`, init: { method: 'GET' } }),
  version: (base) => ({ url: `${base}/admin/version`, init: { method: 'GET' } }),
  query: (base, { sql, params }) => ({
    url: `${base}/query`,
    init: {
      method: 'POST',
      body: JSON.stringify(
        Array.isArray(params) ? { query: sql, params } : { query: sql },
      ),
    },
  }),
  insert: (base, { collection, payload }) => ({
    url: `${base}/collections/${encodeURIComponent(collection)}/rows`,
    init: { method: 'POST', body: JSON.stringify(payload) },
  }),
  bulk_insert: (base, { collection, payloads }) => ({
    url: `${base}/collections/${encodeURIComponent(collection)}/bulk/rows`,
    init: { method: 'POST', body: JSON.stringify({ rows: payloads }) },
  }),
  get: (base, { collection, id }) => ({
    url: `${base}/collections/${encodeURIComponent(collection)}/${encodeURIComponent(id)}`,
    init: { method: 'GET' },
  }),
  delete: (base, { collection, id }) => ({
    url: `${base}/collections/${encodeURIComponent(collection)}/${encodeURIComponent(id)}`,
    init: { method: 'DELETE' },
  }),
  'auth.login': (base, { username, password }) => ({
    url: `${base}/auth/login`,
    init: { method: 'POST', body: JSON.stringify({ username, password }) },
  }),
  'auth.whoami': (base) => ({
    url: `${base}/auth/whoami`,
    init: { method: 'GET' },
  }),
  'auth.create_api_key': (base, params = {}) => ({
    url: `${base}/auth/api-keys`,
    init: { method: 'POST', body: JSON.stringify(params) },
  }),
  'auth.revoke_api_key': (base, { key }) => ({
    url: `${base}/auth/api-keys/${encodeURIComponent(key)}`,
    init: { method: 'DELETE' },
  }),
  'auth.change_password': (base, { current_password, new_password }) => ({
    url: `${base}/auth/change-password`,
    init: {
      method: 'POST',
      body: JSON.stringify({
        current_password,
        new_password,
      }),
    },
  }),
  'cache.get': (base, { namespace, key }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}/${encodeURIComponent(key)}`,
    init: { method: 'GET' },
  }),
  'cache.put': (base, { namespace, key, value, ttl_ms, tags, policy }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}/${encodeURIComponent(key)}`,
    init: { method: 'PUT', body: JSON.stringify({ value, ttl_ms, tags, policy }) },
  }),
  'cache.exists': (base, { namespace, key }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}/${encodeURIComponent(key)}/exists`,
    init: { method: 'GET' },
  }),
  'cache.invalidate': (base, { namespace, key }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}/${encodeURIComponent(key)}`,
    init: { method: 'DELETE' },
  }),
  'cache.invalidate_prefix': (base, { namespace, prefix }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}?prefix=${encodeURIComponent(prefix)}`,
    init: { method: 'DELETE' },
  }),
  'cache.invalidate_tags': (base, { namespace, tags }) => ({
    url: `${base}/cache/ns/${encodeURIComponent(namespace)}/tags`,
    init: { method: 'DELETE', body: JSON.stringify({ tags }) },
  }),
  'cache.flush_namespace': (base, { namespace }) => ({
    url: `${base}/admin/blob_cache/flush_namespace`,
    init: { method: 'POST', body: JSON.stringify({ namespace }) },
  }),
}
